run_md.py

```diff
@@ -0,0 +1,879 @@
+        except subprocess.CalledProcessError as e:
+            logging.error(f'{wdir_md_cur}\t{e}\n')
+            return None
+    else:
+        logging.warning(f'{wdir_md_cur}. Prepared solv_ions.gro file exists. Skip solvation and ion preparation step\n')
+
+    if not prepare_mdp_files(wdir_md_cur=wdir_md_cur, all_resids=all_resids,
+                      script_path=script_path, nvt_time_ps=nvt_time_ps,
+                      npt_time_ps=npt_time_ps, mdtime_ns=mdtime_ns):
+        return None
+
+    return wdir_md_cur
+
+
+def edit_topology_file(topol_file, pattern, add, how='before', n=0):
+    with open(topol_file) as input:
+        data = input.read()
+
+    if n == 0:
+        data = data.replace(pattern, f'{add}\n{pattern}' if how == 'before' else f'{pattern}\n{add}')
+    else:
+        data = data.split('\n')
+        ind = data.index(pattern)
+        data.insert(ind+n, add)
+        data = '\n'.join(data)
+
+    with open(topol_file, 'w') as output:
+        output.write(data)
+
+def get_n_line_of_protein_in_mol_section(topol_file):
+    with open(topol_file) as input:
+        data = input.read()
+    molecules_section = data[data.find('[ molecules ]'):]
+    last_protein_n_line = None
+    for n, line in enumerate(molecules_section.split('\n'), 1):
+        if line.startswith('Protein'):
+            last_protein_n_line = n
+    return last_protein_n_line
+
+
+def add_ligands_to_topol(all_itp_list, all_posres_list, all_resids, topol):
+    itp_include_list, posres_include_list, resid_include_list = [], [], []
+    for itp, posres, resid in zip(all_itp_list, all_posres_list, all_resids):
+        itp_include_list.append(f'; Include {resid} topology\n'
+                                f'#include "{itp}"\n')
+        posres_include_list.append(f'; {resid} position restraints\n#ifdef POSRES_{resid}\n'
+                                  f'#include "{posres}"\n#endif\n')
+        resid_include_list.append(f'{resid}             1')
+
+    edit_topology_file(topol, pattern="; Include forcefield parameters",
+                add='\n'.join(itp_include_list), how='after', n=3)
+    #reverse order since add before pattern
+    edit_topology_file(topol, pattern="; Include topology for ions",
+                add='\n'.join(posres_include_list[::-1]), how='before')
+    # ligand molecule should be after protein (or all protein chains listed)
+    n_line_after_molecules_section = get_n_line_of_protein_in_mol_section(topol_file=topol)
+    edit_topology_file(topol, pattern='[ molecules ]', add='\n'.join(resid_include_list),
+                       how='after', n=n_line_after_molecules_section)
+
+
+def prep_ligand(mol, script_path, project_dir, wdir_ligand, addH=True):
+    molid = mol.GetProp('_Name')
+    resid = mol.GetProp('resid')
+
+    wdir_ligand_cur = os.path.join(wdir_ligand, molid)
+    os.makedirs(wdir_ligand_cur, exist_ok=True)
+
+    mol_file = os.path.join(wdir_ligand_cur, f'{molid}.mol')
```

Comment:
it is better to initialize vars closer to the usage, I suggest to move the line ten lines below